File: hamersapp/src/main/java/nl/ecci/hamers/events/EventListAdapter.java
package nl.ecci.hamers.events;

import android.content.Context;
import android.content.Intent;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Filter;
import android.widget.Filterable;
import android.widget.ImageView;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.Date;

import nl.ecci.hamers.MainActivity;
import nl.ecci.hamers.R;

import static nl.ecci.hamers.helpers.Utils.getOwnUser;

class EventListAdapter extends RecyclerView.Adapter<EventListAdapter.ViewHolder> implements Filterable {

    private final Context context;
    private final ArrayList<Event> dataSet;
    private ArrayList<Event> filteredDataSet;

    EventListAdapter(Context context, ArrayList<Event> dataSet) {
        this.context = context;
        this.dataSet = dataSet;
        this.filteredDataSet = dataSet;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        final View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.event_row, parent, false);
        final ViewHolder vh = new ViewHolder(view);

        view.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                final int position = vh.getAdapterPosition();
                if (position != RecyclerView.NO_POSITION) {
                    Intent intent = new Intent(context, SingleEventActivity.class);
                    intent.putExtra(Event.Companion.getEVENT(), filteredDataSet.get(vh.getAdapterPosition()).getId());
                    context.startActivity(intent);
                }
            }
        });
        return vh;
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        holder.title.setText(filteredDataSet.get(position).getTitle());
        holder.description.setText(filteredDataSet.get(position).getDescription());

        Date date = filteredDataSet.get(position).getDate();
        holder.date.setText(MainActivity.appDF2.format(date));

        if (filteredDataSet.get(position).getLocation() == null || filteredDataSet.get(position).getLocation().equals("null")) {
            holder.location.setVisibility(View.GONE);
        } else {

            holder.location.setText(filteredDataSet.get(position).getLocation());
        }

        ArrayList signups = filteredDataSet.get(position).getSignUps();
        if (signups != null) {
            int userID = getOwnUser(MainActivity.prefs).getId();
            Boolean aanwezig = null;
            for (int i = 0; i < signups.size(); i++) {
                Event.SignUp signUp = (Event.SignUp) signups.get(i);

                if (signUp.getUserID() == userID) {
                    aanwezig = signUp.isAttending();
                }
            }

            if (aanwezig != null) {
                if (aanwezig) {
                    holder.thumbs.setImageResource(R.drawable.ic_thumbs_up);
                } else {
                    holder.thumbs.setImageResource(R.drawable.ic_thumbs_down);
                }
            } else {
                holder.thumbs.setImageResource(R.drawable.ic_questionmark);
            }
        }
    }

    @Override
    public int getItemCount() {
        return filteredDataSet.size();
    }

    @Override
    public Filter getFilter() {
        return new Filter() {
            @Override
            protected FilterResults performFiltering(CharSequence charSequence) {
                FilterResults results = new FilterResults();

                //If there's nothing to filter on, return the original data for your list
                if (charSequence == null || charSequence.length() == 0) {
                    results.values = dataSet;
                    results.count = dataSet.size();
                } else {
                    ArrayList<Event> filterResultsData = new ArrayList<>();
                    for (Event event : dataSet) {
                        if (event.getTitle().toLowerCase().contains(charSequence) || event.getDescription().toLowerCase().contains(charSequence)) {
                            filterResultsData.add(event);
                        }
                    }
                    results.values = filterResultsData;
                    results.count = filterResultsData.size();
                }
                return results;
            }

            @Override
            @SuppressWarnings("unchecked")
            protected void publishResults(CharSequence charSequence, FilterResults filterResults) {
                filteredDataSet = (ArrayList<Event>) filterResults.values;
                notifyDataSetChanged();
            }
        };
    }

    class ViewHolder extends RecyclerView.ViewHolder {
        public final View view;
        public final TextView title;
        public final TextView description;
        public final TextView date;
        final TextView location;
        final ImageView thumbs;

        ViewHolder(View view) {
            super(view);
            this.view = view;

            title = (TextView) view.findViewById(R.id.event_title);
            description = (TextView) view.findViewById(R.id.event_beschrijving);
            date = (TextView) view.findViewById(R.id.event_date);
            location = (TextView) view.findViewById(R.id.event_location);
            thumbs = (ImageView) view.findViewById(R.id.thumbs);
        }
    }
}

File: hamersapp/src/main/java/nl/ecci/hamers/events/EventListAdapter.kt
package nl.ecci.hamers.events

import android.content.Context
import android.content.Intent
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Filter
import android.widget.Filterable
import kotlinx.android.synthetic.main.row_event.view.*
import nl.ecci.hamers.MainActivity
import nl.ecci.hamers.R
import nl.ecci.hamers.helpers.DataUtils
import java.util.*

internal class EventListAdapter(private val context: Context, private val dataSet: ArrayList<Event>) : RecyclerView.Adapter<EventListAdapter.ViewHolder>(), Filterable {

    private var filteredDataSet: ArrayList<Event> = dataSet

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.row_event, parent, false)
        val vh = ViewHolder(view)

        view.setOnClickListener {
            val position = vh.adapterPosition
            if (position != RecyclerView.NO_POSITION) {
                val intent = Intent(context, SingleEventActivity::class.java)
                intent.putExtra(Event.EVENT, filteredDataSet[vh.adapterPosition].id)
                context.startActivity(intent)
            }
        }
        return vh
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bindEvent(filteredDataSet[position])
    }

    override fun getItemCount(): Int {
        return filteredDataSet.size
    }

    override fun getFilter(): Filter {
        return object : Filter() {
            override fun performFiltering(charSequence: CharSequence?): Filter.FilterResults {
                val results = Filter.FilterResults()

                //If there's nothing to filter on, return the original data for your list
                if (charSequence == null || charSequence.isEmpty()) {
                    results.values = dataSet
                    results.count = dataSet.size
                } else {
                    val filterResultsData = dataSet.filterTo(ArrayList<Event>()) { it.title.toLowerCase().contains(charSequence) || it.description.toLowerCase().contains(charSequence) }
                    results.values = filterResultsData
                    results.count = filterResultsData.size
                }
                return results
            }

            override fun publishResults(charSequence: CharSequence, filterResults: Filter.FilterResults) {
                filteredDataSet = filterResults.values as ArrayList<Event>
                notifyDataSetChanged()
            }
        }
    }

    internal inner class ViewHolder(val view: View) : RecyclerView.ViewHolder(view) {

        fun bindEvent(event: Event) {
            with(event) {
                itemView.event_title.text = event.title
                var description = event.description
                if (description.length > 256)
                    description = description.substring(0, 256)
                itemView.event_beschrijving.text = description
                itemView.event_date.text = MainActivity.appDTF.format(event.date)
                if (event.location.isNullOrEmpty()) {
                    itemView.event_location.visibility = View.GONE
                } else {
                    itemView.event_location.text = event.location
                }
            }

            val signUps = event.signUps
            val userID = DataUtils.getOwnUser(context).id
            var aanwezig: Boolean? = null
            signUps.indices
                    .map { signUps[it] }
                    .filter { it.userID == userID }
                    .forEach { aanwezig = it.isAttending }

            if (aanwezig != null) {
                if (aanwezig as Boolean) {
                    itemView.thumbs.setImageResource(R.drawable.ic_thumbs_up)
                } else {
                    itemView.thumbs.setImageResource(R.drawable.ic_thumbs_down)
                }
            } else {
                itemView.thumbs.setImageResource(R.drawable.ic_questionmark)
            }
        }
    }
}
